controllers/controlplane/controlplane.go

```diff
@@ -113,11 +113,12 @@ func (r *CritControlPlaneReconciler) reconcileControlPlane(
 		logger.Info("Initializing control plane", "Replicas", ccp.Spec.Replicas, "Existing", len(ownedMachines))
 		bootstrapSpec := ccp.Spec.CritConfigSpec.DeepCopy()
 
-		// TODO(chrism): replace with NFailureDomainsWithFewestMachines so that
 		// created control plane will be spread out over failure domains
-		fd := controlPlane.FailureDomainWithFewestMachines()
+		fds := controlPlane.NFailureDomainsWithFewestMachines(ccp.Spec.Replicas)
```

Comment:
What are the cases where this would return `nil` or zero-length? Do we need a check here to prevent a panic inside the loop?

Reply:
My original thought was that it would always return a slice of size N, where the values were like nil or empty or something like that, but that was just my naive approach.  Not sure if it is the best way tbh.

Reply:
how does this look?
```
		fds := controlPlane.NFailureDomainsWithFewestMachines(ccp.Spec.Replicas)
		for i := 0; i < ccp.Spec.Replicas; i++ {
			var fd string
			if n := len(fds); n > 0 {
				fd = fds[i%n]
			}

			//I believe this can take a blank string if no fds are present
			if err := r.cloneConfigsAndGenerateMachine(ctx, cluster, ccp, bootstrapSpec, &fd); err != nil { ...
                         ...

```

Reply:
I was thinking something more like this:

```go
// PickNFewest returns N failure domains with the fewest number of machines.
func PickNFewest(failureDomains clusterv1.FailureDomains, machines FilterableMachineCollection, n int) []*string {
	fds := make([]*string, n)
	aggregations := pick(failureDomains, machines)
	if len(aggregations) == 0 {
		return fds
	}
	sort.Sort(aggregations)
	for i := 0; i < n; i++ {
		fds[i] = pointer.StringPtr(aggregations[i%len(aggregations)].id)
	}
	return fds
}
```
where it would return a slice that is always the size of the value `n` that is passed in, that way the caller can indicate how big the slice being constructed should be at the call site.  This also means that you don't have to do a modulo on the length of `fds` in this for loop, you can just directly index it with something like:

```go
		fds := controlPlane.NFailureDomainsWithFewestMachines(ccp.Spec.Replicas)
		for i := 0; i < ccp.Spec.Replicas; i++ {
			if err := r.cloneConfigsAndGenerateMachine(ctx, cluster, ccp, bootstrapSpec, fd[i]); err != nil {
```
The other plus here is that it more mimics the existing function this is adapting in the case that there be no failure domains by having `nil`.

Reply:
I like this idea. I'm assuming for consistency, we would then change: 
```
// NFailureDomainWithFewestMachines returns N failure domains with the fewest number of machines.
// Used to distribute control plane replicas over several failure domains
func (c *ControlPlane) NFailureDomainsWithFewestMachines(replicas int) []*string {
	if len(c.Cluster.Status.FailureDomains.FilterControlPlane()) == 0 {
		return nil
	}
	return PickNFewest(c.FailureDomains().FilterControlPlane(), c.Machines, replicas)
}
```
and make it look like this in order to return a slice that is always the size of the value `n` that is passed in:

```
// NFailureDomainWithFewestMachines returns N failure domains with the fewest number of machines.
// Used to distribute control plane replicas over several failure domains
func (c *ControlPlane) NFailureDomainsWithFewestMachines(replicas int) []*string {
	return PickNFewest(c.FailureDomains().FilterControlPlane(), c.Machines, replicas)
}
```